### PR TITLE
Move trigger_build and handle_commit to WebHookMapping

### PR DESCRIPTION
**Pull request by keto**
The existing trigger_build() method in WebHookMapping class was broken,
because some other parts had been removed. And in the end it seems like
most consistent thing to move eveything out of tasks module.

### CODE REVIEW

File: src/webhook_launcher/app/models.py [hunk 1]
```diff
@@ -287,9 +288,15 @@ def revision(self):
...
-            return _lsr[0]
+        # TODO: refactor the WebHookMapping and LastSeenRevision relation
+        if not hasattr(self, '_lsr'):
+            if self.pk:
+                self._lsr, _ = LastSeenRevision.objects.get_or_create(
+                    mapping=self
+                )
+            else:
+                return None
+        return self._lsr
```
Comment (larstiq) [inline src/webhook_launcher/app/models.py]: Hmm, does this change the caching characteristics to not take a newer lsr if the database changes after object creation?

File: src/webhook_launcher/app/models.py [hunk 2]
```diff
@@ -392,24 +399,119 @@ def clean(self, exclude=None):
...
+    def trigger_build(self, user=None, tag=None, force=False):
+
+        # Only fire for projects which allow webhooks. We can't just
+        # rely on validation since a Project may forbid hooks after
+        # the hook was created
+        if self.project_disabled:
+            print "Project has build disabled"
+            return
+
+        handled = self.lsr.handled and self.lsr.tag == tag and not force
```
Comment (larstiq) [inline src/webhook_launcher/app/models.py]: self.lsr can be None here, hence the LastSeenRevision.objects.get_or_create I imagine

File: src/webhook_launcher/app/models.py [hunk 3]
```diff
@@ -392,24 +399,119 @@ def clean(self, exclude=None):
...
+        if handled:
+            print "build already handled, skipping"
+        build = self.build and self.mapped and not handled
```
Comment (larstiq) [inline src/webhook_launcher/app/models.py]: Does forcing now unset the handled?

File: src/webhook_launcher/app/models.py (hunk 1 above)
Comment (keto) [inline src/webhook_launcher/app/models.py]: Well it changes slightly. But I think it does not make any difference as the life cycle of a WHM object is not that long. 

There probably is chance for race conditions where one process is handling one webhook event when another comes in and ones changes get overwriten, but there haven't really been anything preventing that from happening earlier either... Except that if we run only one participant for handling the webhooks, then I think the events are queued.

This change was mainly to generalize the LastSeenRevision.objects.get_or_create(mapping=mapobj) that is used pretty much every where (which also shows that the FK relation from LSR to WHM is bit backwards, or it shoud be one2one relation)

File: src/webhook_launcher/app/models.py (hunk 2 above)
Comment (keto) [inline src/webhook_launcher/app/models.py]: In theory yes, but in practice trigger_build is always called on saved WHM object, so the lsr property will return a LSR object. But could add a check there that the LSR object is valid.

File: src/webhook_launcher/app/models.py (hunk 3 above)
Comment (keto) [inline src/webhook_launcher/app/models.py]: Yes, forece=True makes handled=False

File: src/webhook_launcher/app/models.py (hunk 1 above)
Comment (larstiq) [inline src/webhook_launcher/app/models.py]: It is something to keep in mind for my django console usage, where the WHM objects are quite long lived.

Fair enough.

File: src/webhook_launcher/app/models.py (hunk 3 above)
Comment (larstiq) [inline src/webhook_launcher/app/models.py]: ok

File: src/webhook_launcher/app/models.py (hunk 2 above)
Comment (keto) [inline src/webhook_launcher/app/models.py]: Added check that the WebHookMapping is saved before calling trigger_build or handle_commit. This ensures that it gets a LastSeenRevision object via lsr property. Whether the lsr has proper values can affect the build result, but checking that is out of scope for this task. For that need to think about the different use cases and what we want to allow or not.

File: src/webhook_launcher/app/models.py [hunk 4]
```diff
@@ -392,24 +399,127 @@ def clean(self, exclude=None):
...
+        qp = None
+        if user is None:
+            user = self.user.username
+
+        if build:
+            if tag:
+                self.lsr.tag = tag
```
Comment (lbt) [inline src/webhook_launcher/app/models.py]: I'm not sure that we should be overwriting the 'tag' in an LSR?
Should that not be immutable once it's come from the payload?

Comment (keto) [inline src/webhook_launcher/app/models.py]: Payload handling does not update the tag, it just passes it to trigger_build.

**Comment by lbt** [comment]
LGTM then
